actioncable/lib/action_cable/connection/web_socket_rack.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module ActionCable
+  module Connection
+    # Wrap the real socket to minimize the externally-presented API
+    #
+    # This variation leverages the `rack.upgrade?` approach detailed here: https://github.com/rack/rack/pull/1272
+    class WebSocketRack # :nodoc:
+      UPGRADE_EXISTS = "rack.upgrade?"
+      UPGRADE = "rack.upgrade"
+      PROTOCOL_NAME_IN = "HTTP_SEC_WEBSOCKET_PROTOCOL"
+      PROTOCOL_NAME_OUT = "Sec-Websocket-Protocol"
+      CLOSE_REASON = ""
+
+      attr_reader :protocol
+
+      def initialize(env, event_target, event_loop, protocols)
+        env[UPGRADE] = self
+        @event_target = event_target
+        @protocol = nil
+        @websocket = nil
+        if env[PROTOCOL_NAME_IN]
+          if env[PROTOCOL_NAME_IN].is_a?(String)   # for single list headers such as "json, soap, foo"
+            env[PROTOCOL_NAME_IN].split(/,[\s]?/).each do |i|
+              next unless protocols.include?(i)
+              @protocol = i
+              break
+            end
+          elsif env[PROTOCOL_NAME_IN].is_a?(Array) # for multiple headers such as: "soap" , "json", "foo"
+            env[PROTOCOL_NAME_IN].each do |i|
+              next unless protocols.include?(i)
+              @protocol = i
+              break
+            end
+          end
+        end
+      end
+
+      def possible?
+        true
+      end
+
+      def alive?
+        websocket&.open?
+      end
+
+      def transmit(data)
+        return false unless websocket
+        case data
+        when Numeric then websocket.write(data.to_s)
+        when String  then websocket.write(data)
+        when Array   then websocket.write(data.pack("C*"))
+        else false
+        end
+      end
+
+      def close
+        websocket&.close
+      end
+
+      def rack_response
+        return [101, { PROTOCOL_NAME_OUT => @protocol }, []] if @protocol
```

Comment:
`@protocol` -> `protocol`?

Reply:
Sure, it's practically the same and now that we're using an `attr_reader` it looks nicer.